src/hist/basehist.py

```diff
@@ -150,20 +161,140 @@ def project(self, *args: Union[int, str]):
         elif all(isinstance(x, str) for x in args):
             indices: tuple = tuple()
             for name in args:
-                for index, axis in enumerate(self.axes):
-                    if name == axis.name:
-                        indices += (index,)
-                        break
-                else:
-                    raise ValueError("The axis names could not be found")
+                if isinstance(
+                    name, str
+                ):  # ToDo: redundant line - mypy cannot realized it
+                    indices += (self._name_to_index(name),)
 
             return super().project(*indices)
 
         else:
             raise TypeError(
-                f"Only projections by indices and names are supported for {self.__class__.__name__}"
+                f"Only projection by indices or names is supported for {self.__class__.__name__}"
+            )
+
+    def fill(self, *args, **kwargs):
```

Comment:
Suggestion: try putting keyword only arguments sample, weight, and thread there, then you don't have to worry about them in the dict, and the signature is nicer for users.

```suggestion
    def fill(self, *args, weight=None, sample=None, thread: Optional[int] =None, **kwargs):
```

Reply:
(And obviously pass them on to the fill's below)